harness/determined/common/storage/shared.py

```diff
@@ -87,13 +101,72 @@ def delete(self, tgt: str) -> None:
 
     def upload(self, src: Union[str, os.PathLike], dst: str) -> None:
         src = os.fspath(src)
-        shutil.copytree(src, os.path.join(self._base_path, dst))
+        python_version = sys.version_info
+
+        if python_version.major == 3 and python_version.minor <= 7:
+            # Workaround for Python3.7 not having dirs_exist_ok for shutil.copytree.
+            _shutil.copytree(src, os.path.join(self._base_path, dst), dirs_exist_ok=True)
+        else:
+            shutil.copytree(src, os.path.join(self._base_path, dst), dirs_exist_ok=True)
```

Comment:
this is a nit, but I think it'd be preferable to do something like this:

```
import shutil

if python_version.major == 3 and python_version.minor <= 7:
  def _copytree(...):
    ...
  def copytree(...):
    ...
else:
  copytree = shutil.copytree

...

class SharedFS(storage.StorageManager):
  ...
  def upload(...):
    ...
    copytree(...)
  
```

One advantage is you are treating the function to call as a variable, rather than using if statements to pass identical args to one of two separate functions.

Another advantage is you have more of the hack "all in one place".  Not completely, since you'd normally not call `copytree(...)`, you'd call `shutil.copytree(...)` instead.  But it's still better I think.

Lastly, the vendored code is literally only even executed in python 3.7.  The way you currently have it written, that `_shutil` modules is getting imported in all versions of python.